app/gui/src/project-view/components/LazyList.vue

```diff
@@ -0,0 +1,250 @@
+<script setup lang="ts" generic="T">
+/**
+ *  @file A list component which instantiates visible elements only.
+ *
+ * The element provided by default slot is used for rendering elements. Only visible elements
+ * are instantiated. See usage in ComponentList as an example.
+ */
+
+import { listBindings } from '@/bindings'
+import { useApproach } from '@/composables/animation'
+import { useResizeObserver } from '@/composables/events'
+import { cloneVNode, computed, h, ref, VNode, watch } from 'vue'
+
+const selected = defineModel<number | null>('selected', { required: false, default: null })
+const {
+  items,
+  itemHeight,
+  scrollToSelectionMargin = 0.0,
+  autoSelectFirst = false,
+  debounceMouseSelection,
+} = defineProps<{
+  /** List of models for an item. */
+  items: readonly T[]
+  itemHeight: number
+  /** A margin kept when scrolling while navigating with arrows */
+  scrollToSelectionMargin?: number
+  /** When set to true, on every `list` prop update first element will be selecetd (if any) */
+  autoSelectFirst?: boolean
+  /**
+   * Debounce time for selecting with mouse hover, expressed in milliseconds.
+   *
+   * Navigating with arrows will still update `selected` model immediately.
+   */
+  debounceMouseSelection?: number
+}>()
+const emit = defineEmits<{
+  itemAccepted: [item: T, index: number]
+}>()
+const slots = defineSlots<{
+  default(props: { item: T }): any
+}>()
+
+const scroller = ref<HTMLElement>()
+
+// === Components List and Positions ===
+
+const visibleComponents = computed(() => {
+  if (scroller.value == null) return []
+  const scrollPos = scrollPosition.value
+  const topmostVisible = itemAtY(scrollPos)
+  const bottommostVisible = Math.max(0, itemAtY(scrollPos + scrollerSize.value.y))
+  return items.slice(topmostVisible, bottommostVisible + 1).map((item, i) => {
+    return { item, index: i + topmostVisible }
+  })
+})
+
+function createVNodes(slot: typeof slots.default) {
+  if (!slot) return undefined
+  return visibleComponents.value.map(({ item, index }) => {
+    return {
+      node: h(
+        'div',
+        { class: 'item', style: itemStyle(index) },
+        slot({ item }).map((node: VNode<unknown, unknown>) =>
+          cloneVNode(node, { class: { selected: index === highlighted.value } }),
+        ),
+      ),
+      item,
+      index,
+    }
+  })
+}
+
+const nodes = computed(() => createVNodes(slots.default))
+
+function itemPos(index: number) {
+  return index * itemHeight
+}
+
+function itemAtY(pos: number) {
+  return Math.floor(pos / itemHeight)
+}
+
+function itemStyle(index: number) {
+  // TODO[ao]: for some reason, .item class style must be repeated here,
```

Comment:
Render functions don't have access to scoped styles. A `<style module>` would work.